File: backend/app/models/ooda.py
# Copyright 2026 Athena Contributors
#
# Use of this software is governed by the Business Source License 1.1
# included in the LICENSE file.
#
# Change Date: Four years from release date of each version
# Change License: Apache License, Version 2.0
#
# For commercial licensing, contact: [TODO: contact email]

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from .enums import OODAPhase


class OODAIteration(BaseModel):
    id: str
    operation_id: str
    iteration_number: int
    phase: OODAPhase
    observe_summary: str | None = None
    orient_summary: str | None = None
    decide_summary: str | None = None
    act_summary: str | None = None
    recommendation_id: str | None = None
    technique_execution_id: str | None = None
    started_at: datetime
    completed_at: datetime | None = None


class OodaTriggerQueued(BaseModel):
    status: Literal["queued"] = "queued"
    operation_id: str


class OODADirectiveCreate(BaseModel):
    directive: str = Field(..., min_length=1, max_length=2000)
    scope: Literal["next_cycle"] = "next_cycle"


class OodaDashboardResponse(BaseModel):
    current_phase: str
    iteration_count: int
    latest_iteration: OODAIteration | None = None
    recent_iterations: list[OODAIteration] = []


class OODADirective(BaseModel):
    id: str
    operation_id: str
    directive: str
    scope: str
    created_at: datetime
    consumed_at: datetime | None = None

File: frontend/src/types/ooda.ts
// Copyright 2026 Athena Contributors
//
// Use of this software is governed by the Business Source License 1.1
// included in the LICENSE file.
//
// Change Date: Four years from release date of each version
// Change License: Apache License, Version 2.0
//
// For commercial licensing, contact: azz093093.830330@gmail.com

import { OODAPhase } from "./enums";

export interface OODAIteration {
  id: string;
  operationId: string;
  iterationNumber: number;
  phase: OODAPhase;
  observeSummary: string | null;
  orientSummary: string | null;
  decideSummary: string | null;
  actSummary: string | null;
  recommendationId: string | null;
  techniqueExecutionId: string | null;
  startedAt: string;
  completedAt: string | null;
}

export interface PhaseDetail {
  factsCount?: number;
  facts?: Array<{ trait: string; value: string; category: string }>;
  rawSummary?: string;
  situationAssessment?: string;
  recommendedTechniqueId?: string;
  confidence?: number;
  reasoningText?: string;
  options?: Array<{
    techniqueId: string;
    techniqueName: string;
    reasoning: string;
    riskLevel: string;
    recommendedEngine: string;
    confidence: number;
    prerequisites: string[];
  }>;
  reason?: string;
  confidenceBreakdown?: Record<string, number>;
  noiseLevel?: string;
  riskLevel?: string;
  matrixAction?: string;
  techniqueId?: string;
  engine?: string;
  status?: string;
  resultSummary?: string;
  errorMessage?: string;
  factsCollectedCount?: number;
  failureCategory?: string;
}

export interface OODATimelineEntry {
  iterationNumber: number;
  phase: string;
  summary: string;
  timestamp: string;
  targetId?: string;
  targetHostname?: string;
  targetIp?: string;
  detail?: PhaseDetail;
}
